Port linear gradients to be interned primitives.

gw3583 opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/webrender/3352)
<!-- Reviewable:end -->

gw3583 commented:
Try run looks good so far. Also seems to be about another 2% win in dl_mutate (https://treeherder.mozilla.org/perf.html#/graphs?series=try,1660472,1,1&selected=try,1660472,407537,652574374).

bors-servo commented:
:umbrella: The latest upstream changes (presumably #3349) made this pull request unmergeable. Please resolve the merge conflicts.

gw3583 commented:
Rebased and fixed one of the review comments. I agree with the other two suggestions too - although would prefer to do those as follow ups, to avoid rebase issues with other open patches.
